File: bumperbot_ws/src/bumperbot_cpp_examples/src/simple_publisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

#include <chrono>

using namespace std::chrono_literals;

class SimplePublisher: public rclcpp::Node
{
public:
  SimplePublisher(): Node("simple_publisher"), counter(0) {
    pub = create_publisher<std_msgs::msg::String>("/chatter", 10);
    //std::chrono::seconds(1)
    timer = create_wall_timer(1s, std::bind(&SimplePublisher::timerCallback, this));

    RCLCPP_INFO(get_logger(), "publishing at 1 Hz");
  }
private:
  unsigned int counter;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub;
  rclcpp::TimerBase::SharedPtr timer;

  void timerCallback() {
    auto message = std_msgs::msg::String();
    message.data = "Hello ROS 2 - counter: " + std::to_string(counter++);

    pub->publish(message);
  }
};

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SimplePublisher>();
  rclcpp::spin(node);
  // shared_ptr이라 따로 free 안해도 됨
  rclcpp::shutdown();

  return 0;
}

File: bumperbot_ws/src/bumperbot_py_examples/bumperbot_py_examples/simple_publisher.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

class SimplePublisher(Node):
  def __init__(self):
    # name of this node is simple_publisher
    super().__init__("simple_publisher")
    # msg_type 송수신할 데이터 타입
    # topic 토픽 이름
    # qos_profile 메시지 큐 사이즈
    self.pub = self.create_publisher(msg_type=String, topic="chatter",qos_profile=10)
    self.counter = 0
    self.frequency = 1.0 # 메시지 게시 속도

    self.get_logger().info(f"publishing at {self.frequency} Hz")

    # 타이머 생성
    self.timer = self.create_timer(self.frequency, self.timerCallback)

  def timerCallback(self):
    msg = String()
    msg.data = f"hello, ros2 - counter = {self.counter}"

    self.pub.publish(msg = msg)
    self.counter += 1


def main():
  rclpy.init() # rclpy 시작. 노드 초기화, 통신 인프라 / ROS 매개변수 / 로깅 / 글로벌 리소스 / 명령 인자 처리 등
  simple_publisher = SimplePublisher() 
  rclpy.spin(simple_publisher) # 연속적으로 노드에 대한 이벤트 처리 수행
  simple_publisher.destroy_node() # 노드를 제거
  rclpy.shutdown() # ros2와 관련된 모든 리소스 정리

if __name__ == "__main__":
  main()